File: test/javascript/erlang/binary_to_float_test.mjs
"use strict";

import {
  assert,
  assertBoxedError,
  defineGlobalErlangAndElixirModules,
} from "../support/helpers.mjs";

import Bitstring from "../../../assets/js/bitstring.mjs";
import Erlang_Binary_To_Float from "../../../assets/js/erlang/binary_to_float.mjs";
import Interpreter from "../../../assets/js/interpreter.mjs";
import Type from "../../../assets/js/type.mjs";

defineGlobalErlangAndElixirModules();

describe("Erlang_Binary_To_Float", () => {
  describe("binary_to_float/1", () => {
    const binary_to_float = Erlang_Binary_To_Float["binary_to_float/1"];

    // ------------------------------
    // Success cases
    // ------------------------------
    it("converts a correct binary to a float", () => {
      const result = binary_to_float(Type.bitstring("10.5"));

      assert.isTrue(Type.isFloat(result));
      assert.strictEqual(result.value, 10.5);
    });

    it("parses scientific notation", () => {
      const result = binary_to_float(Type.bitstring("2.2017764e+1"));

      assert.isTrue(Type.isFloat(result));
      assert.strictEqual(result.value, 22.017764);
    });

    it("parses negative float", () => {
      const result = binary_to_float(Type.bitstring("-3.14"));

      assert.isTrue(Type.isFloat(result));
      assert.strictEqual(result.value, -3.14);
    });

    it("parses float with leading zeros", () => {
      const r = binary_to_float(Type.bitstring("00012.34"));
      assert.strictEqual(r.value, 12.34);
    });

    it("parses + sign float", () => {
      const r = binary_to_float(Type.bitstring("+15.5"));
      assert.strictEqual(r.value, 15.5);
    });

    it("parses uppercase E scientific notation", () => {
      const r = binary_to_float(Type.bitstring("1.2E3"));
      assert.strictEqual(r.value, 1200.0);
    });

    it("parses negative exponent", () => {
      const r = binary_to_float(Type.bitstring("1.23e-3"));
      assert.strictEqual(r.value, 0.00123);
    });

    it("parses negative zero", () => {
      const result = binary_to_float(Type.bitstring("-0.0"));
      assert.isTrue(Interpreter.isEqual(result, Type.float(-0.0)));
    });

    // ------------------------------
    // Error Cases
    // ------------------------------
    it("raises argument error if input is not a binary", () => {
      const term = Type.integer(123);

      assertBoxedError(
        () => binary_to_float(term),
        "ArgumentError",
        "errors were found at the given arguments:\n\n  * 1st argument: expected a binary\n"
      );
    });

    it("raises badarg when float contains underscores", () => {
      const bin = Type.bitstring("1_000.5");

      assertBoxedError(() => binary_to_float(bin), "ErlangError", "badarg");
    });

    it("raises badarg for invalid float format", () => {
      const bin = Type.bitstring("12.3.4");

      assertBoxedError(() => binary_to_float(bin), "ErlangError", "badarg");
    });

    it("raises badarg for non-numeric text", () => {
      const bin = Type.bitstring("abc");

      assertBoxedError(() => binary_to_float(bin), "ErlangError", "badarg");
    });

    it("rejects empty binary", () => {
      assertBoxedError(
        () => binary_to_float(Type.bitstring("")),
        "ErlangError",
        "badarg"
      );
    });

    it("rejects decimal point only", () => {
      assertBoxedError(
        () => binary_to_float(Type.bitstring(".")),
        "ErlangError",
        "badarg"
      );
    });

    it("rejects leading dot such as .5", () => {
      assertBoxedError(
        () => binary_to_float(Type.bitstring(".5")),
        "ErlangError",
        "badarg"
      );
    });

    it("rejects trailing dot such as 5.", () => {
      assertBoxedError(
        () => binary_to_float(Type.bitstring("5.")),
        "ErlangError",
        "badarg"
      );
    });

    it("rejects scientific notation without a fraction like 3e10", () => {
      assertBoxedError(
        () => binary_to_float(Type.bitstring("3e10")),
        "ErlangError",
        "badarg"
      );
    });

    it("raises badarg on trailing exponent marker", () => {
      assertBoxedError(
        () => binary_to_float(Type.bitstring("2e")),
        "ErlangError",
        "badarg"
      );
    });

    it("raises badarg on whitespace around the number", () => {
      assertBoxedError(
        () => binary_to_float(Type.bitstring(" 12.3")),
        "ErlangError",
        "badarg"
      );
    });

    it("raises badarg on multiple exponent markers", () => {
      assertBoxedError(
        () => binary_to_float(Type.bitstring("1e2e3")),
        "ErlangError",
        "badarg"
      );
    });

    it("raises badarg on Infinity text", () => {
      assertBoxedError(
        () => binary_to_float(Type.bitstring("Infinity")),
        "ErlangError",
        "badarg"
      );
    });

    it("raises badarg on hex-style JS float", () => {
      assertBoxedError(
        () => binary_to_float(Type.bitstring("0x1.fp2")),
        "ErlangError",
        "badarg"
      );
    });

    it("rejects non-UTF-8 binary", () => {
      const bin = Bitstring.fromBits([0xff, 0xff, 0xff]);

      assertBoxedError(
        () => binary_to_float(bin),
        "ArgumentError",
        "errors were found at the given arguments:\n\n  * 1st argument: expected a binary\n"
      );
    });
  });
});

File: test/elixir/hologram/ex_js_consistency/erlang/binary_to_float_test.exs
defmodule Hologram.ExJsConsistency.Erlang.BinaryToFloatTest do
  use Hologram.Test.BasicCase, async: true
  @moduletag :consistency

  describe "binary_to_float/1" do
    # -------------------------
    # Success Cases
    # -------------------------

    test "converts a correct binary to a float" do
      result = :erlang.binary_to_float("10.5")
      assert result == 10.5
    end

    test "parses scientific notation" do
      result = :erlang.binary_to_float("2.2017764e+1")
      assert result == 22.017764
    end

    test "parses negative float" do
      result = :erlang.binary_to_float("-3.14")
      assert result == -3.14
    end

    test "parses float with leading zeros" do
      result = :erlang.binary_to_float("00012.34")
      assert result == 12.34
    end

    test "parses + sign float" do
      result = :erlang.binary_to_float("+15.5")
      assert result == 15.5
    end

    test "parses uppercase E scientific notation" do
      result = :erlang.binary_to_float("1.2E3")
      assert result == 1200.0
    end

    test "parses negative exponent" do
      result = :erlang.binary_to_float("1.23e-3")
      assert result == 0.00123
    end

    # -------------------------
    # Error Cases
    # -------------------------

    test "raises argument error if input is not a binary" do
      assert_raise ArgumentError, fn ->
        :erlang.binary_to_float(123)
      end
    end

    test "raises badarg when float contains underscores" do
      assert_raise ArgumentError, fn ->
        :erlang.binary_to_float("1_000.5")
      end
    end

    test "raises badarg for invalid float format" do
      assert_raise ArgumentError, fn ->
        :erlang.binary_to_float("12.3.4")
      end
    end

    test "raises badarg for non-numeric text" do
      assert_raise ArgumentError, fn ->
        :erlang.binary_to_float("abc")
      end
    end

    test "rejects empty binary" do
      assert_raise ArgumentError, fn ->
        :erlang.binary_to_float("")
      end
    end

    test "rejects decimal point only" do
      assert_raise ArgumentError, fn ->
        :erlang.binary_to_float(".")
      end
    end

    test "rejects leading dot such as .5" do
      assert_raise ArgumentError, fn ->
        :erlang.binary_to_float(".5")
      end
    end

    test "rejects trailing dot such as 5." do
      assert_raise ArgumentError, fn ->
        :erlang.binary_to_float("5.")
      end
    end

    test "rejects scientific notation without a fraction like 3e10" do
      assert_raise ArgumentError, fn ->
        :erlang.binary_to_float("3e10")
      end
    end

    test "raises badarg on trailing exponent marker" do
      assert_raise ArgumentError, fn ->
        :erlang.binary_to_float("2e")
      end
    end

    test "raises badarg on whitespace around the number" do
      assert_raise ArgumentError, fn ->
        :erlang.binary_to_float(" 12.3")
      end
    end

    test "raises badarg on multiple exponent markers" do
      assert_raise ArgumentError, fn ->
        :erlang.binary_to_float("1e2e3")
      end
    end

    test "raises badarg on Infinity text" do
      assert_raise ArgumentError, fn ->
        :erlang.binary_to_float("Infinity")
      end
    end

    test "raises badarg on hex-style JS float" do
      assert_raise ArgumentError, fn ->
        :erlang.binary_to_float("0x1.fp2")
      end
    end
  end
end
